src/lib/core/datetime.c

```diff
@@ -259,3 +260,154 @@ datetime_compare(const struct datetime *lhs, const struct datetime *rhs)
 
 	return COMPARE_RESULT(lhs->nsec, rhs->nsec);
 }
+
+/**
+ * Interval support functions: stringization and operations
+ */
+
+#define NANOS_PER_SEC 1000000000
+
+static inline int64_t
+interval_seconds(const struct interval *ival)
+{
+	return (int64_t)ival->sec + ival->nsec / NANOS_PER_SEC;
+}
+
+static inline int
+interval_days(const struct interval *ival)
+{
+	return interval_seconds(ival) / SECS_PER_DAY;
+}
+
+static inline int
+interval_hours(const struct interval *ival)
+{
+	return interval_seconds(ival) / (60 * 60);
+}
+
+static inline int
+interval_minutes(const struct interval *ival)
+{
+	return interval_seconds(ival) / 60;
+}
+
+/**
+ * If |nsec| is larger than allowed range 1e9 then modify passed
+ * sec accordingly.
+ */
+static void
+denormalize_interval_nsec(int64_t *psec, int *pnsec)
+{
+	assert(psec != NULL);
+	assert(pnsec != NULL);
+	int64_t sec = *psec;
+	int nsec = *pnsec;
+	/*
+	 * There is nothing to change:
+	 * - if there is a small nsec with 0 in sec;
+	 * - or if both sec and nsec have the same sign, and nsec is
+	 *   small enough.
+	 */
+	if (nsec == 0)
+		return;
+
+	bool zero_or_same_sign = sec == 0 || (sec < 0) == (nsec < 0);
+	if (zero_or_same_sign && labs(nsec) < NANOS_PER_SEC)
+		return;
+
+	int64_t normalized_nsec = sec * NANOS_PER_SEC + nsec;
+	sec = normalized_nsec / NANOS_PER_SEC;
+	nsec = normalized_nsec % NANOS_PER_SEC;
+
+	*psec = sec;
+	*pnsec = nsec;
+}
+
+/**
+ * SNPRINT-like function to generate textual representation of a sec.nsec
+ * given buffer and size of buffer
+ */
+static size_t
+seconds_str(char *buf, ssize_t len, bool need_sign, long sec, int nsec)
+{
+	sec %= 60;
+	denormalize_interval_nsec(&sec, &nsec);
+	size_t sz = 0;
+	bool is_neg = sec < 0 || (sec == 0 && nsec < 0);
+
+	SNPRINT(sz, snprintf, buf, len, is_neg ? "-" : (need_sign ? "+" : ""));
+	sec = labs(sec);
+	if (nsec != 0) {
+		SNPRINT(sz, snprintf, buf, len, "%" PRId64 ".%09ld seconds",
+			sec, labs(nsec));
+	} else {
+		SNPRINT(sz, snprintf, buf, len, "%" PRId64 " seconds", sec);
+	}
+	return sz;
+}
+
+#define SPACE() \
+	do { \
+		if (sz > 0) { \
+			SNPRINT(sz, snprintf, buf, len, ", "); \
+		} \
+	} while (0)
+
+size_t
+interval_to_string(const struct interval *ival, char *buf, ssize_t len)
```

Comment:
Yup, as we still try to output seconds and nanoseconds in human-readable form (to simplify cognitive efforts necessary for converting nanoseconds into fractions of seconds). This may be considered as questionable, but a bit of ergonomics should not be a problem, IMVHO.

Though, the boundaries against which we are checking nanoseconds are `MIN_INT32`/`MAX_INT32` which imply ~+-2e9. We may enforce strict 1e9 for a moment (till leap seconds to come) and such normalization would not bring us more than 1 seconds in each direction.

Reply:
My comment was rather related to this inconsistency:
when it comes to nsec, you convert them to seconds, once the nsec part overflows 1e9.
when it comes to seconds, you don't convert them to minutes, hence the example with "+1161 seconds" instead of "+19 minutes 21 seconds"
Is this intended?

Reply:
Also shouldn't `dt.interval.new{min=1, sec=59, nsec=2e9+1}` be printed like "+2 minutes, 1.000000001 seconds" ?

Reply:
> My comment was rather related to this inconsistency:
> when it comes to nsec, you convert them to seconds, once the nsec part overflows 1e9.
> when it comes to seconds, you don't convert them to minutes, hence the example with "+1161 seconds" instead of "+19 minutes 21 seconds"
> Is this intended?
> Also shouldn't `dt.interval.new{min=1, sec=59, nsec=2e9+1}` be printed like "+2 minutes, 1.000000001 seconds" ?

Sigh. This is once again and again about this attempt to output intervals in more readable form (with these exercises to `denormalize_nsec` upon output), but still be in bounds of original RFC which required to keep all fields separate, to make their operations be dependent only on final date we will apply them.

I'd rather not go the way to recursively "normalize" minutes, hours and such. (Because they would violate RFC and produce output not representing actual data from interval object)

So, the longer we go, the more intentions to get rid of this denormalization of nanoseconds (making it more consistent with other fields, but less human readable). Let me verify how quick this change (regression) might be done...

Reply:
Created #7045 so it might be taken care separately.

Reply:
Therefore, I propose to write nanoseconds separately from seconds. Still, there is a difference between "one minute = 60 seconds" and "one second = 1000000000 nanoseconds" because the former is defined as a ratio between time intervals, while the latter is just a value and a "nano"-value (which is 10^-9). Relationships between time intervals seem a little looser than "nano".